nslsii/temperature_controllers.py

```diff
@@ -35,29 +37,44 @@ def __init__(self, pv_prefix, **kwargs):
     # define the new set method with the new moving indicator
     def set(self, value):
         # check that a set is not in progress, and if not set the lock.
-        if self._set_lock:
-            raise Exception('attempting to set {}'.format(self.name) +
+        if not self._set_lock.acquire(blocking=False):
+            raise Exception('attempting to set {} '.format(self.name) +
                             'while a set is in progress'.format(self.name))
-        self._set_lock = True
 
         # define some required values
         set_value = value
-        status = DeviceStatus(self, timeout=self.timeout.get())
+        status = DeviceStatus(self)
 
         initial_timestamp = None
 
         # grab these values here to avoidmutliple calls.
         equilibrium_time = self.equilibrium_time.get()
         tolerance = self.tolerance.get()
 
+        # setup a cleanup function for the timer, this matches including
+        # timeout in `status` but also ensures that the callback us removed.
+        def timer_cleanup(status):
+            print('Set of {} timed out after {} s'.format(self.name,
+                                                          self.timeout.get()))
+            self._set_lock.release()
+            self.readback.clear_sub(status_indicator)
+            status._finished(success=False)
+
+        cb_timer = th.Timer(self.timeout.get(), timer_cleanup, args=[status])
```

Comment:
The `[status]` jumped out at me. It's a minor thing, but it's safer to use a tuple `(status,)` where possible and basically always for listing positional arguments

Reply:
done